compiler/src/dotty/tools/dotc/util/NameTransformer.scala

```diff
@@ -55,6 +55,29 @@ object NameTransformer {
     else name
   }
 
+  /** Decode expanded characters starting with `$u`, followed by the character's unicode expansion. */
+  def decodeIllegalChars(name: String): String = {
+    if (name.contains("$u")) {
+      val sb = new mutable.StringBuilder()
+      var i = 0
+      while (i < name.length - 5) {
+        if (name(i) == '$' && name(i + 1) == 'u') {
+          sb.append(Integer.valueOf(name.substring(i + 2, i + 6), 16).toChar)
+          i += 6
+        } else {
+          sb.append(name(i))
+          i += 1
+        }
+      }
+      while (i < name.length) {
+        sb.append(name(i))
+        i += 1
+      }
+      sb.result()
+    }
+    else name
+  }
```

Comment:
Is `decodeIllegalChars` meant to invert `avoidIllegalChars`? Ideally, a few tests `decodeIllegalChars(avoidIllegalChars(str)) == str` would be good.

Potential, shorter alternative for the 2 while loops:
```scala
      while (i < name.length) {
        if (i < name.length - 5 && name(i) == '$' && name(i + 1) == 'u') {
          sb.append(Integer.valueOf(name.substring(i + 2, i + 6), 16).toChar)
          i += 6
        } else {
          sb.append(name(i))
          i += 1
        }
      }
```